kvserver: avoid races where replication changes can get interrupted

aayushshah15 opened the pull request:
This commit adds a safeguard inside
`Replica.maybeLeaveAtomicChangeReplicasAndRemoveLearners()` to avoid removing
learner replicas _when we know_ that that learner replica is in the process of
receiving its initial snapshot (as indicated by an in-memory lock on log
truncations that we place while the snapshot is in-flight).

This change should considerably reduce the instances where `AdminRelocateRange`
calls are interrupted by the mergeQueue or the replicateQueue (and vice versa).

Fixes https://github.com/cockroachdb/cockroach/issues/57129
Relates to https://github.com/cockroachdb/cockroach/issues/79118

Release note: none

Jira issue: CRDB-14769

aayushshah15 committed:
kvserver: avoid races where replication changes can get interrupted

This commit adds a safeguard inside
`Replica.maybeLeaveAtomicChangeReplicasAndRemoveLearners()` to avoid removing
learner replicas _when we know_ that that learner replica is in the process of
receiving its initial snapshot (as indicated by an in-memory lock on log
truncations that we place while the snapshot is in-flight).

This change should considerably reduce the instances where `AdminRelocateRange`
calls are interrupted by the mergeQueue or the replicateQueue (and vice versa).

Fixes https://github.com/cockroachdb/cockroach/issues/57129
Relates to https://github.com/cockroachdb/cockroach/issues/79118

Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/79379)
<!-- Reviewable:end -->

tbg commented on pkg/kv/kvserver/replica_command.go:
```diff
@@ -1289,8 +1289,15 @@ func (r *Replica) maybeLeaveAtomicChangeReplicasAndRemoveLearners(
+	// initial upreplication snapshot, we don't want to remove them and we bail
+	// out.
+	if r.hasOutstandingLearnerSnapshotInFlight() {
+		return desc, errors.New("cannot remove learner while a snapshot is in flight")
```
Nothing wrong with this code, but it'll give us new test flakes.

aayushshah15 replied:
I've made it such that we don't wait out the grace period for snapshot application when this method checks for the presence of an in-flight snapshot. All the flakes I was seeing (thus far) were because of that 3-second grace period.

tbg commented on pkg/kv/kvserver/replica_raft.go:
```diff
@@ -1584,6 +1614,9 @@ func (r *Replica) getAndGCSnapshotLogTruncationConstraintsLocked(
 		if recipientStore != 0 && item.recipientStore != recipientStore {
 			continue
 		}
+		if item.deadline != (time.Time{}) && ignoreDeadline {
```
This seems named the wrong way around. If we ignore the deadline, shouldn't we look at all entries? We're ignoring entries that have a deadline, so consider renaming this param to `ignoreEntsWithDeadline`.
Also add a comment. It's not clear when the deadline is or isn't set. I think a deadline is set only before a transfer is inflight, but I am just guessing (and I wrote this code originally!)

tbg replied:
maybe `onlyInflight` is a better name.

aayushshah15 replied:
Thanks for taking a look! I've improved an existing comment and added another one explaining when the deadline is assigned. Let me know if you think it still needs to be improved.

tbg reviewed (commented):
The code looks good, but I'm still surprised this doesn't cause issues in tests. When before a split would nuke the learner and do its thing, it will now fail with an opaque error that isn't going to be handled anywhere, so it will bubble up to the caller. For example, in `workload kv init --splits 1000`, if the cluster is still up-replicating, the error *has* to pop up at some point, no? And we certainly would have tests that do this. We should definitely make the error here structured so that we can backfill code to handle it where we find it sorely needed. And we may want to retry the error, at least for a little while. In some sense, what we really want is to block, not to fail fast, and I suspect that if we land the change as is, we will find our hand forced in that direction in practice.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/79379#-:-MzvGtwh69DHCQdYUwAt:b-2c84zo)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aayushshah15 and @nvanbenschoten)

___
*[pkg/kv/kvserver/replica_command.go, line 1296 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/79379#-MzvG4X22jgprJURRXpE:-MzvG4X22jgprJURRXpF:bu3caq3) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf92d52afdde7acfa10019d4a96f172c43a2bb50/pkg/kv/kvserver/replica_command.go#L1296)):*
> ```Go
> 	// out.
> 	if r.hasOutstandingLearnerSnapshotInFlight() {
> 		return desc, errors.New("cannot remove learner while a snapshot is in flight")
> ```

make this a sentinel error (i.e. `var errCannotRemoveLearnerWithInflightSnapshot = errors.New("...")` and then the tests can also reference the error instead of hard-coding the string?
___
*[pkg/kv/kvserver/replica_raft.go, line 1539 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/79379#-MzvGTFE0nzTReb4OeEn:-MzvGTFE0nzTReb4OeEo:b-ah3k9b) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf92d52afdde7acfa10019d4a96f172c43a2bb50/pkg/kv/kvserver/replica_raft.go#L1539)):*
> ```Go
> 		index:          index,
> 		recipientStore: recipientStore,
> 		// NB: We explicitly do not want to set a deadline here, since the presence
> ```

Consider a rename of the field

https://github.com/cockroachdb/cockroach/blob/0cc83717e5a32c69cd4917d82a7481fc73e1c402/pkg/kv/kvserver/replica_raft.go#L1564-L1565

to something more evocative of its semantics, like `completedAt`, and move the `now.Add(gracePeriod)` check into the the method that actually removes the entries. This will work better with the new use case introduced in this PR.
___
*[pkg/kv/kvserver/replica_learner_test.go, line 1205 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/79379#-MzvFzrj2zfbeYAnUMNm:-MzvFzrj2zfbeYAnUMNn:b8q2fwj) ([raw file](https://github.com/cockroachdb/cockroach/blob/bb2c29c77e9b28e8f9a81a822ed99d75b03aeb14/pkg/kv/kvserver/replica_learner_test.go#L1205)):*
> ```Go
> 			_ = tc.AddVotersOrFatal(t, scratchStartKey, tc.Target(1))
> 		})
> ```

spurious
___
*[pkg/kv/kvserver/replica_learner_test.go, line 1211 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/79379#-MzvFWie8uWFrB5qVTjj:-MzvFWie8uWFrB5qVTjk:bt03q8) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf92d52afdde7acfa10019d4a96f172c43a2bb50/pkg/kv/kvserver/replica_learner_test.go#L1211)):*
> ```Go
> 	atomic.StoreInt64(&activateSnapshotTestingKnob, 1)
> 	replicationChange := make(chan error)
> 	go func() {
> ```

Please use the stopper here.
___
*[pkg/kv/kvserver/replica_learner_test.go, line 1232 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/79379#-MzvFvH67zo0CkpMlVvz:-MzvFvH67zo0CkpMlVw-:b-fleng7) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf92d52afdde7acfa10019d4a96f172c43a2bb50/pkg/kv/kvserver/replica_learner_test.go#L1232)):*
> ```Go
> 	db.Exec(t, `SET CLUSTER SETTING kv.range_merge.queue_enabled = true`)
> 
> 	testutils.SucceedsWithin(
> ```

Why not just SucceedsSoon?
___
*[pkg/kv/kvserver/replica_learner_test.go, line 1255 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/79379#-MzvFxzK1MZV63mNhQKe:-MzvFxzK1MZV63mNhQKf:b8q2fwj) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf92d52afdde7acfa10019d4a96f172c43a2bb50/pkg/kv/kvserver/replica_learner_test.go#L1255)):*
> ```Go
> 	defer leaktest.AfterTest(t)()
> 	defer log.Scope(t).Close(t)
> ```

spurious


<!-- Sent from Reviewable.io -->

aayushshah15 reviewed (commented):
> When before a split would nuke the learner and do its thing, it will now fail with an opaque error that isn't going to be handled anywhere, so it will bubble up to the caller

It looks like splits only ever cared about exiting joint config states, never about removing learner replicas: https://github.com/cockroachdb/cockroach/blob/7858c2908323c25dcea8083c47c1e459a336f5d5/pkg/kv/kvserver/replica_command.go#L316-L322

>And we may want to retry the error, at least for a little while. In some sense, what we really want is to block, not to fail fast, and I suspect that if we land the change as is, we will find our hand forced in that direction in practice.

At the moment, the only entities that will fail-fast when they encounter an in-flight learner snapshot are the `mergeQueue`, the `StoreRebalancer`, or the `replicateQueue` (if it encounters the range while either the `mergeQueue` or the `StoreRebalancer` are rebalancing it). That seems good, doesn't it? As long as we're not blocking splits, we _should_ fail-fast and let these queues move on to other ranges, right?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/79379#-:-Mzw6C7yD2Sxva3YFdxE:b-dg1n0)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten and @tbg)

___
*[pkg/kv/kvserver/replica_command.go, line 1296 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/79379#-MzvG4X22jgprJURRXpE:-Mzw5FXt82VCN5vE_v0p:bs7wumx) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf92d52afdde7acfa10019d4a96f172c43a2bb50/pkg/kv/kvserver/replica_command.go#L1296)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

make this a sentinel error (i.e. `var errCannotRemoveLearnerWithInflightSnapshot = errors.New("...")` and then the tests can also reference the error instead of hard-coding the string?
</blockquote></details>

Done. 
___
*[pkg/kv/kvserver/replica_raft.go, line 1539 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/79379#-MzvGTFE0nzTReb4OeEn:-Mzw5CqI7Ly7phfhcgY2:b-s9g368) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf92d52afdde7acfa10019d4a96f172c43a2bb50/pkg/kv/kvserver/replica_raft.go#L1539)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Consider a rename of the field

https://github.com/cockroachdb/cockroach/blob/0cc83717e5a32c69cd4917d82a7481fc73e1c402/pkg/kv/kvserver/replica_raft.go#L1564-L1565

to something more evocative of its semantics, like `completedAt`, and move the `now.Add(gracePeriod)` check into the the method that actually removes the entries. This will work better with the new use case introduced in this PR.
</blockquote></details>

Done, much cleaner.
___
*[pkg/kv/kvserver/replica_learner_test.go, line 1205 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/79379#-MzvFzrj2zfbeYAnUMNm:-Mzw5sklCXQQrlE_piVQ:bt2vr1b) ([raw file](https://github.com/cockroachdb/cockroach/blob/bb2c29c77e9b28e8f9a81a822ed99d75b03aeb14/pkg/kv/kvserver/replica_learner_test.go#L1205)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

spurious
</blockquote></details>

Fixed.
___
*[pkg/kv/kvserver/replica_learner_test.go, line 1211 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/79379#-MzvFWie8uWFrB5qVTjj:-Mzw5aeGEOCLdIrdw9jm:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf92d52afdde7acfa10019d4a96f172c43a2bb50/pkg/kv/kvserver/replica_learner_test.go#L1211)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Please use the stopper here.
</blockquote></details>

Done.
___
*[pkg/kv/kvserver/replica_learner_test.go, line 1232 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/79379#-MzvFvH67zo0CkpMlVvz:-Mzw5pU25pXmPhvO-N-z:b-cb1g4d) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf92d52afdde7acfa10019d4a96f172c43a2bb50/pkg/kv/kvserver/replica_learner_test.go#L1232)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Why not just SucceedsSoon?
</blockquote></details>

No particular reason, changed.
___
*[pkg/kv/kvserver/replica_learner_test.go, line 1255 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/79379#-MzvFxzK1MZV63mNhQKe:-Mzw5s-L9RwB41B9cdjG:bt2vr1b) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf92d52afdde7acfa10019d4a96f172c43a2bb50/pkg/kv/kvserver/replica_learner_test.go#L1255)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

spurious
</blockquote></details>

Fixed.


<!-- Sent from Reviewable.io -->

aayushshah15 commented:
> It looks like splits only ever cared about exiting joint config states, never about removing learner replicas

If we feel that this is undesirable and we do want splits to be able to remove learners before they begin, we could override that `hasOutstandingLearnerSnapshotInFlight()` check just for splits. 





<!-- Sent from Reviewable.io -->

tbg reviewed (approved):
> It looks like splits only ever cared about exiting joint config states, never about removing learner replicas

Ah, thanks for reminding me of that. I think that is fine.

What about SCATTER? Hard to avoid the error there, lest we are willing to retry "for a long time", or return with empty hands.

> Reviewed 3 of 5 files at r2, 3 of 3 files at r3, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/79379#-:-N--FOjWC7zWjp8IlJVE:b-9isbdg)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aayushshah15 and @nvanbenschoten)

___
*[pkg/kv/kvserver/replicate_queue.go, line 1255 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/79379#-N--FafiBFxFAIVsiwxv:-N--FafiBFxFAIVsiwxw:b1bgd3l) ([raw file](https://github.com/cockroachdb/cockroach/blob/bb2c29c77e9b28e8f9a81a822ed99d75b03aeb14/pkg/kv/kvserver/replicate_queue.go#L1255)):*
> ```Go
> }
> 
> func (rq *replicateQueue) removeLearner(
> ```

:ribbon: 


<!-- Sent from Reviewable.io -->

aayushshah15 committed:
kvserver: remove detritus related to pre-emptive snapshots

This commit removes the grace period (and associated logic) that the sender of
a snapshot used to adhere to before GC-ing that snapshot's log truncation
constraint. This code was only meant to guard against races where a pre-emptive
snapshot was received by a recipient but not applied (which would happen when
the associated replication change was completed). This race is no longer
possible today.

Release note: None

aayushshah15 reviewed (commented):
> What about SCATTER? 

It seems like scatter hitting those errors should be okay. Scatter doesn't return any rebalancing related errors up to its client, so that's not a concern:
https://github.com/cockroachdb/cockroach/blob/5fc368e81642e254b0aa56b845a65d81b627e3c5/pkg/kv/kvserver/replica_command.go#L3318-L3327

Moreover, if scatter does encounter a range that the `replicateQueue` is already in the process of moving, it seems behaviorally fine for scatter to not also move it.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/79379#-:-N-0e_QHCBklKY3sUJMl:bpsrt03)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aayushshah15, @nvanbenschoten, and @tbg)



<!-- Sent from Reviewable.io -->

aayushshah15 reviewed (commented):
@nvanbenschoten I've swapped these commits around like you mentioned.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/79379#-:-N-43ROh6PITLZKubbJl:bvhmsjj)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aayushshah15, @nvanbenschoten, and @tbg)



<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
> Reviewed 1 of 5 files at r6, 7 of 7 files at r7, 7 of 7 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/79379)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aayushshah15)

___
*[pkg/kv/kvserver/replica_command.go, line 1282 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/79379#-N-6qt6-Dx2iesufVAqE:-N-6qt6-Dx2iesufVAqF:b-gn792) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca1d0725c89975b5ead8ff564d1c67706cb26703/pkg/kv/kvserver/replica_command.go#L1282)):*
> ```Go
> }
> 
> // ErrCannotRemoveLearnerWhileSnapshotInFlight is returned when we cannot remove
> ```

Does this need to be marked as a form of `IsExpectedRelocateError` and of `IsRetriableReplicationChangeError`?
___
*[pkg/kv/kvserver/replica_command.go, line 1295 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/79379#-N-6pIR496XIFIZqQsqx:-N-6pIR496XIFIZqQsqy:b-ylnt9l) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca1d0725c89975b5ead8ff564d1c67706cb26703/pkg/kv/kvserver/replica_command.go#L1295)):*
> ```Go
> 	desc, err = r.maybeLeaveAtomicChangeReplicas(ctx, desc)
> 	if err != nil {
> 		return nil, learnersRemoved, err
> ```

If we know `learnersRemoved` is 0 here, it's clearer to return 0, not the named return arg. For example, this would be even harder to read as `return rangeDesc, learnersRemoved, err`. From that line, it's not possible to determine what is being returned. A reader now needs to understand the entire control flow of this function above this point to determine which cases `learnersRemoved` is non-zero.

I think we should make this change wherever we can in this function.
___
*[pkg/kv/kvserver/replica_command.go, line 1302 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/79379#-N-6swlU-FJc06pcFeOM:-N-6swlU-FJc06pcFeON:bylm1ns) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca1d0725c89975b5ead8ff564d1c67706cb26703/pkg/kv/kvserver/replica_command.go#L1302)):*
> ```Go
> 	// out.
> 	if r.hasOutstandingLearnerSnapshotInFlight() {
> 		return desc, learnersRemoved, ErrCannotRemoveLearnerWhileSnapshotInFlight
> ```

How do we expect each of the possible callers into this function to react to this error? Will all of them react gracefully? It would be worthwhile to enumerate them all and ensure that their retry behavior (or lack thereof) in the presence of this error is desirable.
___
*[pkg/kv/kvserver/replica_raft.go, line 1557 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/79379#-N-6orOr7U_vlAxCfm1E:-N-6orOr7U_vlAxCfm1F:b-os25tf) ([raw file](https://github.com/cockroachdb/cockroach/blob/712ff6d9a9c486ac52b4ddf98e4a4a4c1d2293f9/pkg/kv/kvserver/replica_raft.go#L1557)):*
> ```Go
> 	}
> 
> 	item.completed = true
> ```

Do we need this status anymore? Can we delete from `r.mu.snapshotLogTruncationConstraints` immediately?
___
*[pkg/kv/kvserver/replica_raft.go, line 1602 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/79379#-N-6q6lOAGxPgU055hlh:-N-6q6lOAGxPgU055hli:b-m1fjtg) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca1d0725c89975b5ead8ff564d1c67706cb26703/pkg/kv/kvserver/replica_raft.go#L1602)):*
> ```Go
> 	learners := r.Desc().Replicas().LearnerDescriptors()
> 	for _, repl := range learners {
> 		if yes := r.hasOutstandingSnapshotInFlightToStore(repl.StoreID); yes {
> ```

nit: along the lines of my comment above, the following is slightly easier to read because it preserves less state in local variables and allows readers to read backward.
```go
if r.hasOutstandingSnapshotInFlightToStore(repl.StoreID) {
    return true
}
```
___
*[pkg/kv/kvserver/replicate_queue.go, line 670 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/79379#-N-6s19_0V2H2AoByuWg:-N-6s19a8uCZQdU4fgAD:bmghj23) ([raw file](https://github.com/cockroachdb/cockroach/blob/712ff6d9a9c486ac52b4ddf98e4a4a4c1d2293f9/pkg/kv/kvserver/replicate_queue.go#L670)):*
> ```Go
> 		return rq.removeDead(ctx, repl, deadNonVoterReplicas, nonVoterTarget, dryRun)
> 
> 	case AllocatorRemoveLearner:
> ```

Was this just dead code before?
___
*[pkg/kv/kvserver/replicate_queue.go, line 663 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/79379#-N-6scbsAqh1o8u3-bwc:-N-6scbsAqh1o8u3-bwd:bywy75v) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca1d0725c89975b5ead8ff564d1c67706cb26703/pkg/kv/kvserver/replicate_queue.go#L663)):*
> ```Go
> 		return rq.removeDead(ctx, repl, deadNonVoterReplicas, nonVoterTarget, dryRun)
> 
> 	case AllocatorRemoveLearner:
> ```

Should we combine this with `AllocatorFinalizeAtomicReplicationChange` to avoid redundancy?

```
case AllocatorRemoveLearner, AllocatorFinalizeAtomicReplicationChange:
```
___
*[pkg/kv/kvserver/replicate_queue.go, line 690 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/79379#-N-6sCfU60VtsVPXaMmA:-N-6sCfU60VtsVPXaMmB:b4e94jq) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca1d0725c89975b5ead8ff564d1c67706cb26703/pkg/kv/kvserver/replicate_queue.go#L690)):*
> ```Go
> 		}
> 		rq.metrics.RemoveLearnerReplicaCount.Inc(learnersRemoved)
> 		return true, err
> ```

`return true, nil`


<!-- Sent from Reviewable.io -->

aayushshah15 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/79379)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aayushshah15, @nvanbenschoten, and @tbg)

___
*[pkg/kv/kvserver/replica_command.go, line 1282 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/79379#-N-6qt6-Dx2iesufVAqE:-N-EdCt9CM9HCDVaAuNK:bhrfukn) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca1d0725c89975b5ead8ff564d1c67706cb26703/pkg/kv/kvserver/replica_command.go#L1282)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Does this need to be marked as a form of `IsExpectedRelocateError` and of `IsRetriableReplicationChangeError`?
</blockquote></details>

There don't seem to be any tests that expect to run into a learner replica in the process of receiving its first snapshot, so I don't think it needs to be marked `IsExpectedRelocateError` or marked retriable. 
___
*[pkg/kv/kvserver/replica_command.go, line 1295 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/79379#-N-6pIR496XIFIZqQsqx:-N-EeDxr7LzKKuufUY4B:b-s9g368) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca1d0725c89975b5ead8ff564d1c67706cb26703/pkg/kv/kvserver/replica_command.go#L1295)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

If we know `learnersRemoved` is 0 here, it's clearer to return 0, not the named return arg. For example, this would be even harder to read as `return rangeDesc, learnersRemoved, err`. From that line, it's not possible to determine what is being returned. A reader now needs to understand the entire control flow of this function above this point to determine which cases `learnersRemoved` is non-zero.

I think we should make this change wherever we can in this function.
</blockquote></details>

Done, much cleaner.
___
*[pkg/kv/kvserver/replica_command.go, line 1302 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/79379#-N-6swlU-FJc06pcFeOM:-N-EeJML7ThXU1WSk7Ky:b2xuapu) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca1d0725c89975b5ead8ff564d1c67706cb26703/pkg/kv/kvserver/replica_command.go#L1302)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

How do we expect each of the possible callers into this function to react to this error? Will all of them react gracefully? It would be worthwhile to enumerate them all and ensure that their retry behavior (or lack thereof) in the presence of this error is desirable.
</blockquote></details>

The callers that could potentially run into this error are:
1. mergeQueue
2. StoreRebalancer
3. replicateQueue

In all of these cases, they will fail-fast and retry after their respective intervals (which is 10 mins for the queues, and ~1 min for the StoreRebalancer). It does seem desirable to let these queues move on to other ranges when they encounter this error, as opposed to waiting for a while for the learner to finish receiving its snapshot. Retrying or waiting on this error would make issues like https://github.com/cockroachdb/cockroach/issues/79249 more likely. 
___
*[pkg/kv/kvserver/replica_raft.go, line 1557 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/79379#-N-6orOr7U_vlAxCfm1E:-N-EfibkEMX2Lc_84cv9:b-hevc2a) ([raw file](https://github.com/cockroachdb/cockroach/blob/712ff6d9a9c486ac52b4ddf98e4a4a4c1d2293f9/pkg/kv/kvserver/replica_raft.go#L1557)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Do we need this status anymore? Can we delete from `r.mu.snapshotLogTruncationConstraints` immediately?
</blockquote></details>

I've cleaned it up now.
___
*[pkg/kv/kvserver/replica_raft.go, line 1602 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/79379#-N-6q6lOAGxPgU055hlh:-N-EfoUK-hceTU9dn1xm:b-rb74gs) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca1d0725c89975b5ead8ff564d1c67706cb26703/pkg/kv/kvserver/replica_raft.go#L1602)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: along the lines of my comment above, the following is slightly easier to read because it preserves less state in local variables and allows readers to read backward.
```go
if r.hasOutstandingSnapshotInFlightToStore(repl.StoreID) {
    return true
}
```
</blockquote></details>

You're right, done. 
___
*[pkg/kv/kvserver/replicate_queue.go, line 670 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/79379#-N-6s19_0V2H2AoByuWg:-N-EfqIx9nm1VBpMgUau:bb77ucv) ([raw file](https://github.com/cockroachdb/cockroach/blob/712ff6d9a9c486ac52b4ddf98e4a4a4c1d2293f9/pkg/kv/kvserver/replicate_queue.go#L670)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Was this just dead code before?
</blockquote></details>

Yep.
___
*[pkg/kv/kvserver/replicate_queue.go, line 663 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/79379#-N-6scbsAqh1o8u3-bwc:-N-EfrUj8I44_tfKCASq:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca1d0725c89975b5ead8ff564d1c67706cb26703/pkg/kv/kvserver/replicate_queue.go#L663)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Should we combine this with `AllocatorFinalizeAtomicReplicationChange` to avoid redundancy?

```
case AllocatorRemoveLearner, AllocatorFinalizeAtomicReplicationChange:
```
</blockquote></details>

Done.
___
*[pkg/kv/kvserver/replicate_queue.go, line 690 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/79379#-N-6sCfU60VtsVPXaMmA:-N-EfsOj0xH3V4BIBYsB:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca1d0725c89975b5ead8ff564d1c67706cb26703/pkg/kv/kvserver/replicate_queue.go#L690)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

`return true, nil`
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 9 of 9 files at r9, 7 of 7 files at r10, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/79379#-:-N-OUYzN1r0nRn2-NNZc:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @aayushshah15)

___
*[pkg/kv/kvserver/replica_command.go, line 1282 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/79379#-N-6qt6-Dx2iesufVAqE:-N-OSZSE8wSs-Ehn9Dtl:b-ggwoso) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca1d0725c89975b5ead8ff564d1c67706cb26703/pkg/kv/kvserver/replica_command.go#L1282)):*
<details><summary><i>Previously, aayushshah15 (Aayush Shah) wrote…</i></summary><blockquote>

There don't seem to be any tests that expect to run into a learner replica in the process of receiving its first snapshot, so I don't think it needs to be marked `IsExpectedRelocateError` or marked retriable. 
</blockquote></details>

I'm surprised that `TestKVNemesisMultiNode` doesn't run into this. Do you mind stressing that test for a bit?
___
*[pkg/kv/kvserver/replica_command.go, line 1302 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/79379#-N-6swlU-FJc06pcFeOM:-N-OUAID9xTMIiEpeiSA:bjz3e1o) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca1d0725c89975b5ead8ff564d1c67706cb26703/pkg/kv/kvserver/replica_command.go#L1302)):*
> The callers that could potentially run into this error are:

And some of the SQL-level admin operations that result in `AdminChangeReplicas` and `AdminRelocateRange` requests, right?

> In all of these cases ...

Thanks for explaining. This all checks out. Do you mind adding a comment along these lines here?


<!-- Sent from Reviewable.io -->

aayushshah15 commented:
*[pkg/kv/kvserver/replica_command.go, line 1302 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/79379#-N-6swlU-FJc06pcFeOM:-N-PHG8B99ewZ1NVlyCM:br0hynn) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca1d0725c89975b5ead8ff564d1c67706cb26703/pkg/kv/kvserver/replica_command.go#L1302)):*
>SQL-level admin operations

That's a good catch, I've made it such that we won't fail an entire `EXPERIMENTAL_RELOCATE` statement if it encounters a range that is undergoing a rebalance. Do you mind taking a look at the logic inside `sql/relocate.go`?

>Do you mind adding a comment along these lines here?

Done.


<!-- Sent from Reviewable.io -->

aayushshah15 commented:
*[pkg/kv/kvserver/replica_command.go, line 1282 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/79379#-N-6qt6-Dx2iesufVAqE:-N-PX9Os-C5GCvz4RIqQ:b-fvxkf4) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca1d0725c89975b5ead8ff564d1c67706cb26703/pkg/kv/kvserver/replica_command.go#L1282)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

I'm surprised that `TestKVNemesisMultiNode` doesn't run into this. Do you mind stressing that test for a bit?
</blockquote></details>

KVNemesis was indeed running into the error after about ~1k iterations on the default config, I've now taught it to ignore this error.


<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
> Reviewed 4 of 5 files at r12, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/79379)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @aayushshah15, @nvanbenschoten, and @tbg)

___
*[pkg/kv/kvserver/replica_command.go, line 1313 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/79379#-N-THVo8EoJkav2tiW2v:-N-THVo96ype2YW0BfJO:b6y5wx5) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d572b4ae65e672dfcff9d63770ca0184d453046/pkg/kv/kvserver/replica_command.go#L1313)):*
> ```Go
> 	if r.hasOutstandingLearnerSnapshotInFlight() {
> 		return nil /* desc */, 0, /* learnersRemoved */
> 			errors.Mark(ErrCannotRemoveLearnerWhileSnapshotInFlight, ErrMarkReplicationChangeInProgress)
> ```

Can we mark this statically? (i.e. lift the `errors.Mark` to the global)
___
*[pkg/sql/relocate.go, line 149 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/79379#-N-TbgLm8HXscFwy98vY:-N-TbgLn--_HxC5go2Gm:b-s121n3) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d572b4ae65e672dfcff9d63770ca0184d453046/pkg/sql/relocate.go#L149)):*
> ```Go
> 		return false, errors.AssertionFailedf("unknown relocate mode: %v", n.subjectReplicas)
> 	}
> 	// If the `AdminRelocateRange`call failed because it found that the range was
> ```

nit: missing space before call
___
*[pkg/sql/relocate.go, line 149 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/79379#-N-Tetzt8-uGT5sxZo-d:-N-Tetzt8-uGT5sxZo-e:b-hq1dya) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d572b4ae65e672dfcff9d63770ca0184d453046/pkg/sql/relocate.go#L149)):*
> ```Go
> 		return false, errors.AssertionFailedf("unknown relocate mode: %v", n.subjectReplicas)
> 	}
> 	// If the `AdminRelocateRange`call failed because it found that the range was
> ```

Are we sure we want to ignore these errors in this case? Could this lead to complaints that "RELOCATE doesn't work?" Do we want a retry loop with backoff here?
___
*[pkg/sql/relocate.go, line 149 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/79379#-N-TfWqpCFH1_aLdDrp-:-N-TfWqq6f9PVMTe8ZVM:ba7ddvq) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d572b4ae65e672dfcff9d63770ca0184d453046/pkg/sql/relocate.go#L149)):*
> ```Go
> 		return false, errors.AssertionFailedf("unknown relocate mode: %v", n.subjectReplicas)
> 	}
> 	// If the `AdminRelocateRange`call failed because it found that the range was
> ```

Do we need something like this in `pkg/sql/relocate_range.go` as well?


<!-- Sent from Reviewable.io -->

aayushshah15 commented:
*[pkg/sql/relocate.go, line 149 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/79379#-N-Tetzt8-uGT5sxZo-d:-N-TlRCo7LF_p8zauzue:bfp9klw) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d572b4ae65e672dfcff9d63770ca0184d453046/pkg/sql/relocate.go#L149)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Are we sure we want to ignore these errors in this case? Could this lead to complaints that "RELOCATE doesn't work?" Do we want a retry loop with backoff here?
</blockquote></details>

Yeah, I'm not sure what the right thing to do here is. Backoff seems undesirable compared to just forcing a removal of the learner replica in progress, since we'll likely be backing off to then undo this work anyway. I thought about perhaps adding a parameter to "force" `AdminRelocateRange` and `AdminChangeReplicas` to continue with removing learners if they're issued via SQL, but that has mixed version considerations. Do you think that's worth the complexity (as part of this patch specifically)?

Given all that, I do find myself leaning towards the current approach that ignores these ranges and moves on to other ones, maybe with an eye to fix it in a future patch that force-removes these learners. What do you think?


<!-- Sent from Reviewable.io -->

aayushshah15 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/79379)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @aayushshah15, @nvanbenschoten, and @tbg)

___
*[pkg/kv/kvserver/replica_command.go, line 1313 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/79379#-N-THVo8EoJkav2tiW2v:-N0CLgNcC_oW417SYlVU:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d572b4ae65e672dfcff9d63770ca0184d453046/pkg/kv/kvserver/replica_command.go#L1313)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Can we mark this statically? (i.e. lift the `errors.Mark` to the global)
</blockquote></details>

Done.
___
*[pkg/sql/relocate.go, line 149 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/79379#-N-Tetzt8-uGT5sxZo-d:-N0CKTmeCvnR20PER_ZP:b-2hnz6o) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d572b4ae65e672dfcff9d63770ca0184d453046/pkg/sql/relocate.go#L149)):*
<details><summary><i>Previously, aayushshah15 (Aayush Shah) wrote…</i></summary><blockquote>

Yeah, I'm not sure what the right thing to do here is. Backoff seems undesirable compared to just forcing a removal of the learner replica in progress, since we'll likely be backing off to then undo this work anyway. I thought about perhaps adding a parameter to "force" `AdminRelocateRange` and `AdminChangeReplicas` to continue with removing learners if they're issued via SQL, but that has mixed version considerations. Do you think that's worth the complexity (as part of this patch specifically)?

Given all that, I do find myself leaning towards the current approach that ignores these ranges and moves on to other ones, maybe with an eye to fix it in a future patch that force-removes these learners. What do you think?

</blockquote></details>

OK, I've actually changed this back to returning an error / failing the SQL statement if we run into this error. That seems better, at least in the short term, than silently ignoring ranges because of the concern you outlined. 
___
*[pkg/sql/relocate.go, line 149 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/79379#-N-TbgLm8HXscFwy98vY:-N0CKhJjEwDR-56D_LJK:bt2vr1b) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d572b4ae65e672dfcff9d63770ca0184d453046/pkg/sql/relocate.go#L149)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: missing space before call
</blockquote></details>

Fixed.


<!-- Sent from Reviewable.io -->

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 1 of 12 files at r13, 11 of 12 files at r14, 1 of 1 files at r15, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/79379#-:-N0bNxEW3PDMtGd7ady7:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @aayushshah15)

___
*[`pkg/sql/relocate.go` line 149 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/79379#-N-Tetzt8-uGT5sxZo-d:-N0bNsTb2lm1OO76IF7r:bb2u6oo) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d572b4ae65e672dfcff9d63770ca0184d453046/pkg/sql/relocate.go#L149)):*
<details><summary><i>Previously, aayushshah15 (Aayush Shah) wrote…</i></summary><blockquote>

OK, I've actually changed this back to returning an error / failing the SQL statement if we run into this error. That seems better, at least in the short term, than silently ignoring ranges because of the concern you outlined. 
</blockquote></details>

I agree. That seems like a better short-term solution.


<!-- Sent from Reviewable.io -->

aayushshah15 commented:
TFTR!

bors r+

aayushshah15 commented:
blathers backport 22.1
